transit.md

```diff
@@ -202,6 +237,10 @@ IP address or a domain. Furthermore, IPv6 link-local addresses are not supported
 ```json
 {
   "type": "relay-v1",
+  "name": "<string, optional>",
+  "urls": [
+    "<protocol>://<host>:<port>",
+  ],
```

Comment:
Since we're already communicating structured data here, I really think this would benefit from _not_ trying to follow a single "URL" format (that will need its own parser anyway) and instead look more like:

```
urls: [
    {
        "protocol": "tcp",
        "host": "...",
        "port": 1234
    }
]
```

Reply:
...and maybe `transports` or something is more appropriate than `urls` in that case.

Reply:
I'd be open to changing this to a more structured format, but how would the WebSockets hint look like? WebSockets paths are already URLs by nature and thus rather complex (a WebSocket can be on an arbitrary path).

The other point is that most implementations will need an URL parser anyways, because that's the most natural way to let users input them. (Also, every WebSockets implementation already has a full-blown HTTP client inside, so the dependencies are already there)

Reply:
Okay, quick alternative proposal for the encoding:

- add a `type` field to `hints` to distinguish different transports, where we assume `TCP` as default if omitted for backwards compatibility.
- a new `type: WebSockets` hint is added: `{ "type": "WebSockets", "url": "ws://example.com" }`. It features only an URL that you can directly feed into the WebSockets library without hazzle. In that same spirit, `ws` and `wss` hints use the same kind of hint.
- The `urls` array is not added anymore
- The `url_hints` in the ability needs to get a different name then. (Maybe `fancy_transport`? :D)

Example:

```json

{
  "type": "relay-v1",
  "name": "Example relay",
  "hints": [
    {
      "type": "TCP",
      "hostname": "example.org",
      "port": "1234"
    },
    {
      "type": "WebSockets",
      "url": "wss://example.org:8000"
    }
  ],
}
```

Reply:
I like the look of that a lot better. This would allow expansion into other protocols that may or may not have a well-defined `proto://` style thing (_is_ `tcp:` actually well-defined?) like QUIC, Tor, or others.

For sure websocket transports _do_ involve the whole HTTP stack of course.

I would maybe stick with lower-case only for the type to avoid confusion (which has actually happened in WebSocket things, unfortunately). `"type": "tcp"` or `"type": "websocket"` (and that is defined as "websocket" or "WebSocket" .. i.e. without the "s").

The RFC specs the value as all lower-case when it's a value .. e.g. `Upgrade: websocket` .. and HTTP headers are mixed-case, but IIRC _that_ RFC says they shall be interpreted as case-insensitive.

So for the hints stuff, talking about the version, you meant you prefer the `"url_hints": true` thing to just bumping that to `relay-v2` .. because the actual relay protocol isn't changing, just the hints-exchange method...?

Reply:
> is tcp: actually well-defined

Kinda? I did some research initially and I found a few applications using it that way, but no official "standard" of any sorts.

Your naming proposals are good.

 > So for the hints stuff, talking about the version, you meant you prefer the "url_hints": true thing to just bumping that to relay-v2 .. because the actual relay protocol isn't changing, just the hints-exchange method...?

Yes. Basically, the change is backwards compatible and clients that support the new features can work together with clients that don't pretty easily. (As far as I checked)

Reply:
I'm still kind of split on the version -- having stuff "continue to work" is nice, but the bool flag is .. not nice ;)

Reply:
What would be your preferred way of handling the versioning? We could add `relay-v2` instead and say that they are identical except for the hints encoding, but I haven't thought any of this through yet.

Also I'd like to know what is your problem with adding additional data to the ability? Is it because it would be difficult to map to language or framework constructs in some way? (I always look at these from the Rust perspective, but Rust has very powerful enums so maybe that's unfair)

Reply:
I tried out the `relay-v2` approach in the latest commit, let me know what you think.

Reply:
I don't have a strong objection to adding things to the abilities .. it's just that a bool like that kind of feels like a "different version"..? Like if you have a function `def decrypt(bool is_aes):` it might be better to have two functions `def decrypt_aes()` and `def decrypt_default()` instead.

I do like that latest commit a lot (a couple typos / nitpicks, see below). Lets think about backwards-compatibility in both cases:

*only `relay-v1`*: implementations select on the abilities `url_hints` to enable support for parsing new-version hints. Peers with `relay-v1` ability but either peer having `url_hints: false` will choose non-URL hints and pass only those.

*`relay-v2` and `relay-v1`*: implementations select parsing based on the abilities string. Peers where either one has no `relay-v2` ability will only pass old-style hints.

Should be fairly similar, implementation-wise. I think I prefer the `relay-v2` one, as it seems more explicit.